cranelift/codegen/src/peepmatic.rs

```diff
@@ -0,0 +1,847 @@
+            let inst = pos.func.dfg.value_def(v).inst()?;
+            if pos.func.dfg.inst_results(inst).len() == 1 {
+                Some(inst)
+            } else {
+                None
+            }
+        }
+        Part::Constant(c) => {
+            let v = const_to_value(pos.ins(), c, root);
+            let inst = pos.func.dfg.value_def(v).unwrap_inst();
+            Some(inst)
+        }
+        Part::ConditionCode(_) => None,
+    }
+}
+
+fn part_to_value(pos: &mut FuncCursor, root: Inst, part: Part<ValueOrInst>) -> Option<Value> {
+    match part {
+        Part::Instruction(ValueOrInst::Inst(inst)) => {
+            pos.func.dfg.inst_results(inst).first().copied()
+        }
+        Part::Instruction(ValueOrInst::Value(v)) => Some(v),
+        Part::Constant(c) => Some(const_to_value(pos.ins(), c, root)),
+        Part::ConditionCode(_) => None,
+    }
+}
+
+impl Opcode {
+    fn to_peepmatic_operator(&self) -> Option<Operator> {
+        macro_rules! convert {
+            ( $( $op:ident $(,)* )* ) => {
+                match self {
+                    $( Self::$op => Some(Operator::$op), )*
+                    _ => None,
+                }
+            }
+        }
+
+        convert!(
+            AdjustSpDown,
+            AdjustSpDownImm,
+            Band,
+            BandImm,
+            Bconst,
+            Bint,
+            Bor,
+            BorImm,
+            Brnz,
+            Brz,
+            Bxor,
+            BxorImm,
+            Iadd,
+            IaddImm,
+            Icmp,
+            IcmpImm,
+            Iconst,
+            Ifcmp,
+            IfcmpImm,
+            Imul,
+            ImulImm,
+            Ireduce,
+            IrsubImm,
+            Ishl,
+            IshlImm,
+            Isub,
+            Rotl,
+            RotlImm,
+            Rotr,
+            RotrImm,
+            Sdiv,
+            SdivImm,
+            Select,
+            Sextend,
+            Srem,
+            SremImm,
+            Sshr,
+            SshrImm,
+            Trapnz,
+            Trapz,
+            Udiv,
+            UdivImm,
+            Uextend,
+            Urem,
+            UremImm,
+            Ushr,
+            UshrImm,
+        )
+    }
+}
+
+impl TryFrom<Constant> for Imm64 {
+    type Error = &'static str;
+
+    fn try_from(c: Constant) -> Result<Self, Self::Error> {
+        match c {
+            Constant::Int(x, _) => Ok(Imm64::from(x as i64)),
+            Constant::Bool(..) => Err("cannot create Imm64 from Constant::Bool"),
+        }
+    }
+}
+
+impl Into<Constant> for Imm64 {
+    #[inline]
+    fn into(self) -> Constant {
+        let x: i64 = self.into();
+        Constant::Int(x as _, BitWidth::SixtyFour)
+    }
+}
+
+impl Into<Part<ValueOrInst>> for Imm64 {
+    #[inline]
+    fn into(self) -> Part<ValueOrInst> {
+        let c: Constant = self.into();
+        c.into()
+    }
+}
+
+fn part_to_imm64(pos: &mut FuncCursor, part: Part<ValueOrInst>) -> Imm64 {
```

Comment:
If this is for a signed operation, the immediate needs to be sign extended.

Reply:
I haven't been able to create a test case where I can observe the lack of sign extension, and the fact that the existing code wasn't doing this either makes me think it isn't necessary. Can you provide a test case that shows this?

Reply:
See https://github.com/bytecodealliance/wasmtime/issues/1095

Reply:
Thanks. I'll work on fixing #1095 at the `InstructionBuilder` level, as suggested in that issue, rather than in this PR.